parsers/ZA.py

```diff
@@ -0,0 +1,964 @@
+        # ---------------------------------------
+        # if last_cleansed_csv_len != curr_cleansed_csv_len:
+        #     print(f"{i}/{len(merged_csv_bi_data)}: {key}")
+        #     print(f"last: {last_cleansed_csv_len} curr: {curr_cleansed_csv_len}")
+        #     pp.pprint(value)
+        #     print("-----------------------------------------------------")
+
+        last_cleansed_csv_len = curr_cleansed_csv_len
+        cleansed_csv_value_updates.append(curr_update_dict)
+
+    # Mapping columns to keys
+    # Helpful: https://www.eskom.co.za/dataportal/glossary/
+    column_mapping = {
+        0: "coal",  # Thermal_Gen_Excl_Pumping_and_SCO
+        1: "unknown",  # Eskom_OCGT_SCO_Pumping             changed to unknown since negative oil is not possible
+        2: "unknown",  # Eskom_Gas_SCO_Pumping              changed to unknown since negative gas is not possible
+        3: "hydro",  # Hydro_Water_SCO_Pumping
+        4: "hydro",  # Pumped_Water_SCO_Pumping
+        5: "unknown",  # Thermal_Generation                 sum of 0, 1, 2, 3, 4. This value is not used.
+        6: "nuclear",  # Nuclear_Generation
+        7: "unknown",  # International_Imports
+        8: "oil",  # Eskom_OCGT_Generation
+        9: "gas",  # Eskom_Gas_Generation
+        10: "oil",  # Dispatchable_IPP_OCGT
+        11: "hydro",  # Hydro_Water_Generation
+        12: "hydro",  # Pumped_Water_Generation
+        13: "unknown",  # IOS_Excl_ILS_and_MLR              Interruption of Supply. Not used.
+        14: "unknown",  # ILS_Usage                         Interruptible Load Shed = companies paid not to consume electricity. Not used.
+        15: "unknown",  # Manual_Load_Reduction_MLR         MLS = forced load shedding. Not used.
+        16: "wind",  # Wind
+        17: "solar",  # PV
+        18: "solar",  # CSP
+        19: "unknown",  # Other_RE
+    }
+
+    all_data = []
+
+    for i, (datetime_key, value) in enumerate(merged_csv_bi_data.items()):
+
+        # The production values (MW) should never be negative. Use None, or omit the key if # a specific production mode is not known.
+        # ---
+        # storage values can be both positive (when storing energy) or negative (when the
+        # storage is discharged).
+
+        data = {
+            "zoneKey": zone_key,
+            "datetime": datetime_key.datetime,
+            "production": {
+                "biomass": None,
+                "coal": 0.0,
+                "gas": 0.0,
+                "nuclear": 0.0,
+                "oil": 0.0,
+                "solar": 0.0,
+                "wind": 0.0,
+                "geothermal": None,
+                "unknown": 0.0,
+            },
+            "storage": {"hydro": 0.0},
+            "source": " https://www.eskom.co.za",
+        }
+
+        cleansed_csv_values = value["cleansed_csv"].split(",")
+
+        storage_inversion_idcs = [3, 4, 11, 12]
+        production_idcs = [0, 1, 2, 6, 8, 9, 10, 16, 17, 18, 19]
```

Comment:
Amazing job guys! 🤩. Just a small nitpick

At [Eskom Glossary](https://www.eskom.co.za/dataportal/glossary/) they make the difference between  

- Hydro Generation – Generation from large hydro power stations, and sent-out onto the Transmission network.

- Pumped Water Generation – Generation from pumped storage power stations, and sent-out onto the Transmission network.

That will also go in line with the historical data we got from the TSO.

So I suggest including `# Hydro_Water_Generation` in production. 

```suggestion
        storage_inversion_idcs = [3, 4, 12]
        production_idcs = [0, 1, 2, 6, 8, 9, 10, 11, 16, 17, 18, 19]
```

Reply:
Ah ok, tbh I find their notation quite confusing. 

We have the following four hydro values:
```
3: "hydro",  # Hydro_Water_SCO_Pumping
4: "hydro",  # Pumped_Water_SCO_Pumping
---
11: "hydro",  # Hydro_Water_Generation
12: "hydro",  # Pumped_Water_Generation
```

As far as I understood, 4 and 12 are complimentary (Pumped_Water_SCO_Pumping vs. Pumped_Water_Generation). Given your explanation for 11 still keeps me confused regarding 3. What is meant by "Hydro_Water_SCO_Pumping"?

Reply:
In the light of SCO = Synchronous Condenser Operation, and the way they use pumping for non-charging things (that is, pumped hydro; there are no batteries in the data IIRC),  I strongly suspect that in these cases they use 'pumping' for power plant self-consumption. Basically the power drawn from the grid for keeping the generators spinning and the lights on even if the plant isn't generating any power itself.

IIRC @electricitymap policy was to ignore self-consumption data. (Arguably it should be included in total consumption, but that raises the problem of how to account for it when a power plant *is* generating power and provides for its own self-consumption, so the power draw for that isn't reported because it never 'enters' or 'leaves' the grid. Therefore,  it makes sense to ignore it altogether for the sake of consistency.)